python/kvikio/kvikio/remote_file.py

```diff
@@ -0,0 +1,151 @@
+        """S3 context, which initializes and maintains the S3 SDK and client.
+
+        The S3Context calls `Aws::InitAPI()` and `Aws::ShutdownAPI`, which inherit
+        some limitations from the SDK: please construct and destruct `S3Context`
+        from the same thread (use a dedicated thread if necessary). This avoids
+        problems in initializing the dependent Common RunTime C libraries.
+
+        Please make sure that AWS credentials have been configured on the system.
+        A common way to do this, is to set the environment variables:
+        `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.
+
+        Other relevant options are `AWS_DEFAULT_REGION` and `AWS_ENDPOINT_URL`, see
+        <https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html>.
+
+        Parameters
+        ----------
+        endpoint_override
+            If not empty, the address of the S3 server. This takes precedences over
+            the AWS system configuration including the `AWS_ENDPOINT_URL` environment
+            variable.
+        """
+        self._handle = _get_remote_module().S3Context(endpoint_override)
+
+
+class RemoteFile:
+    """File handle of a remote file (currently, only AWS S3 is supported)."""
+
+    def __init__(self, context: S3Context, bucket_name: str, object_name: str):
+        """Open a remote file given a bucket and object name.
+
+        Parameters
+        ----------
+        context
+            The S3 context used for the connection to the remove server.
+        bucket_name
+            Name of the bucket.
+        object_name
+            Name of the object.
+        """
+        self._handle = _get_remote_module().RemoteFile.from_bucket_and_object(
+            context._handle, bucket_name, object_name
+        )
+
+    @classmethod
+    def from_url(cls, context: S3Context, url: str) -> RemoteFile:
+        """Open a remote file given an url such as "s3://<bucket>/<object>".
+
+        Parameters
+        ----------
+        context
+            The S3 context used for the connection to the remove server.
+        url
+            URL to the remote file.
+
+        Returns
+        -------
+        A newly opened remote file
+        """
+        ret = object.__new__(cls)
+        ret._handle = _get_remote_module().RemoteFile.from_url(context._handle, url)
+        return ret
+
+    def __enter__(self) -> RemoteFile:
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        pass
+
+    def nbytes(self) -> int:
+        """Get the file size.
+
+        Note, this is very fast, no communication needed.
+
+        Returns
+        -------
+        The number of bytes.
+        """
+        return self._handle.nbytes()
+
+    def read(self, buf, size: Optional[int] = None, file_offset: int = 0) -> int:
+        """Read from remote source into buffer (host or device memory) in parallel.
+
+        Parameters
+        ----------
+        buf: buffer-like or array-like
```

Comment:
Missing a space before the type. See https://numpydoc.readthedocs.io/en/latest/format.html#parameters.
```suggestion
        buf : buffer-like or array-like
```